### PR TITLE
merge-cachi2-sboms: re-generate test data, fix uncovered bugs

### PR DESCRIPTION
**Pull request by chmeliik**
https://issues.redhat.com/browse/STONEBLD-3045

Primarily, the need to do this came from working on SPDX. I needed to get equivalent SPDX SBOMs for the CycloneDX SBOMs in our test data. But nobody knows how those were generated, so I re-generated them in a more repeatable way. And that uncovered some bugs.

See the individual commits for more details

### CODE REVIEW

**Comment by chmeliik** [comment]
Reviewer guide:
* first commit (the vast majority of lines changed) is a script for generating test data + generated test data
* everything up to the `handle ...` commits is refactoring
* the two `handle ...` commits are bugfixes

**Comment by chmeliik** [comment]
I noticed the generated SBOMs were changing every time I reran the `test_data/assemble-input-sboms.sh` script. Fixed that.

File: sbom-utility-scripts/scripts/merge-cachi2-sboms-script/merge_cachi2_sboms.py [hunk 1]
```diff
@@ -30,26 +80,29 @@ def _is_cachi2_non_registry_dependency(component: dict) -> bool:
...
     This is done by taking a purl and removing any qualifiers and subpaths.
 
     See https://github.com/package-url/purl-spec/tree/master#purl for more info on purls.
     """
-    url = urlsplit(component["purl"])
-    return url.scheme + ":" + url.path
+    purl = component.purl()
+    if not purl:
+        raise ValueError(f"cachi2 component with no purl? name={component.name()}, version={component.version()}")
+    return purl._replace(qualifiers=None, subpath=None).to_string()
```
Comment (chmeliik) [inline sbom-utility-scripts/scripts/merge-cachi2-sboms-script/merge_cachi2_sboms.py:102]: Just FYI: NamedTuple methods don't follow the same naming scheme as most other python classes. Despite starting with an underscore, the `_replace` method is public API https://docs.python.org/3/library/collections.html#collections.somenamedtuple._replace

(PackageURL is a NamedTuple)

File: sbom-utility-scripts/scripts/merge-cachi2-sboms-script/test_merge_cachi2_sboms.py [hunk 2]
```diff
@@ -44,6 +48,42 @@ def test_merge_sboms(data_dir: Path) -> None:
 
     assert json.loads(result) == expected_sbom
 
+    with open(f"{data_dir}/cachi2.bom.json") as f:
+        cachi2_sbom = json.load(f)
+
+    purls_taken_from_syft_sbom = get_purls(expected_sbom) - get_purls(cachi2_sbom)
+    assert sorted(purls_taken_from_syft_sbom) == [
```
Comment (tnevrlka) [inline sbom-utility-scripts/scripts/merge-cachi2-sboms-script/test_merge_cachi2_sboms.py:55]: Does the `purls_taken_from_syft_sbom` need to be sorted? Could the compared hardcoded list be a set instead?

**Review by tnevrlka — COMMENTED** [review]
Looks great to me, super tiny nitpicks. 
I have zero previous experience with this though, so you might want another reviewer to take a look too

File: sbom-utility-scripts/scripts/merge-cachi2-sboms-script/merge_cachi2_sboms.py [hunk 3]
```diff
@@ -3,13 +3,22 @@
 from argparse import ArgumentParser
 from dataclasses import dataclass
 from typing import Any, Callable, Protocol, Sequence
-from urllib.parse import quote_plus, urlsplit
+from urllib.parse import quote_plus
+
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply sbom-utility-scripts/scripts/merge-cachi2-sboms-script/merge_cachi2_sboms.py:8]: Between `urllib.parse` and `packageurl`? Yes, the convention is to separate stdlib imports from third-party imports. Some python formatters automatically separate them like that, e.g. [isort](https://pycqa.github.io/isort/) (or [ruff's isort implementation](https://docs.astral.sh/ruff/rules/#isort-i), which is what I use)

File: sbom-utility-scripts/scripts/merge-cachi2-sboms-script/test_merge_cachi2_sboms.py (hunk 2 above)
Comment (chmeliik) [reply sbom-utility-scripts/scripts/merge-cachi2-sboms-script/test_merge_cachi2_sboms.py]: Yup, changed to a set :+1:

**Comment by chmeliik** [comment]
@brunoapimentel you're probably the only one with more experience regarding this script if you want to take a look :smile:

**Review by tnevrlka — APPROVED** [review]

**Comment by chmeliik** [comment]
Gonna merge this, there are more big changes coming :sweat_smile:

**Pull request merged**